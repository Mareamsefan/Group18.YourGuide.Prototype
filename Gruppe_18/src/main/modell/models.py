from sqlalchemy import Column, String, Integer, DATETIME
import uuid
from Gruppe_18.src.main.database.sql_alchemy import Base


class Account(Base):
    __tablename__ = "User"
    account_id = Column("id", String, primary_key=True)
    username = Column("username", String)
    password = Column("password", String)
    phoneNumber = Column("phoneNumber", String)
    emailAddress = Column("emailAddress", String)

    def __init__(self, username, password, phoneNumber, emailAddress):
        self.account_id = str(uuid.uuid4())
        self.username = username
        self.password = password
        self.phoneNumber = phoneNumber
        self.emailAddress = emailAddress

    def __repr__(self):
        return f"({self.account_id}) {self.username} {self.password} {self.phoneNumber} {self.emailAddress}"

    def get_id(self):
        return self.id


class Tour(Base):
    # duration in hours
    # cost in dollars
    __tablename__ = "Tour"
    tour_id = Column("id", String, primary_key=True)
    title = Column("title", String)
    date = Column("date", DATETIME)
    destination = Column("destination", String)
    duration = Column("duration", Integer)
    cost = Column("cost", Integer)
    max_travelers = Column("max_travelers", Integer)
    language = Column("language", String)
    pictureURL = Column("pictureURL", String)
    booked = Column("booked", Integer, default=0)

    def __init__(self, title, date, destination, duration, cost, max_travelers, language, pictureURL):
        self.tour_id = str(uuid.uuid4())
        self.title = title
        self.date = date
        self.destination = destination
        self.duration = duration
        self.cost = cost
        self.language = language
        self.max_travelers = max_travelers
        self.pictureURL = pictureURL

    def __repr__(self):
        return f"({self.tour_id}) {self.title} {self.destination} {self.duration} {self.cost} {self.language} {self.max_travelers} {self.pictureURL} {self.booked}"

