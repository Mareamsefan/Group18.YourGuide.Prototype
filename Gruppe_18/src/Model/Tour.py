import json


class Tour:
    # duration in hours
    # cost in dollars
    # tour_counter:
    tour_counter = 1

    def __init__(self, destination, duration, cost, pictureURL, language="English", max_travelers=15):
        self.pictureURL = pictureURL
        self.language = language
        self.destination = destination
        self.duration = duration
        self.cost = cost
        self.max_travelers = max_travelers
        self.booked = 0
        self.tour_Id = Tour.tour_counter
        Tour.tour_counter += 1

    def book_tour(self):
        if not self.booked >= self.max_travelers:
            self.booked += 1
            return True
        else:
            return False

    def get_tour_description(self):
        return f"This tour will take you to {self.destination} for {self.duration} hours, and is " \
               f"offered in {self.language}"

    def to_dict(self):
        return {"tour_Id": self.tour_Id,
                "destination": self.destination,
                "cost": self.cost,
                "duration": self.duration,
                "pictureURL": self.pictureURL,
                "language": self.language,
                "max_travelers": self.max_travelers}

    def save_to_stream(self, io_stream):
        try:
            filedata = json.load(io_stream)
        except json.JSONDecodeError:
            filedata = []

        duplicate_found = False
        for item in filedata:
            if item == self.to_dict():
                duplicate_found = True
                break

        if not duplicate_found:
            data = self.to_dict()
            filedata.append(data)

        io_stream.truncate(0)
        io_stream.seek(0)
        json.dump(filedata, io_stream, indent=4)

    # er det riktig måte å overføre klasser på til JSON-format?
    def check_if_tour_is_saved(self):
        try:
            with open("../tour.json", "r") as json_file:
                filedata = json.load(json_file)
        except FileNotFoundError:
            filedata = []

        index_is_saved = None
        for i, item in enumerate(filedata):
            if item["tour_Id"] == self.tour_Id:
                index_is_saved = i
                break

        if index_is_saved is not None:
            return True

        else:
            return False

    '''
        def save_to_json(self, filename):
            try:
                with open(filename, "r") as json_file:
                    filedata = json.load(json_file)
            except FileNotFoundError:
                filedata = []
    
            duplicate_found = False
            for item in filedata:
                if item == self.to_dict():
                    duplicate_found = True
                    break
    
            if not duplicate_found:
                data = self.to_dict()
                filedata.append(data)
    
                with open(filename, "w") as json_file:
                    json.dump(filedata, json_file, indent=4)
        '''
