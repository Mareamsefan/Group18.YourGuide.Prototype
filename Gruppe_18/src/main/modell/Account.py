import json
import uuid


class Account:

    def __init__(self, username, password, phoneNumber, emailAddress):
        self.username = username
        self.password = password
        self.phoneNumber = phoneNumber
        self.emailAddress = emailAddress
        self.account_id = str(uuid.uuid4())
        self.registered_tours = []

    def register_to_tour(self, tour):
        if tour.booked < tour.max_travelers:
            tour.book_tour()
            self.registered_tours.append(tour)
        else:
            return "you can't register to this tour"
