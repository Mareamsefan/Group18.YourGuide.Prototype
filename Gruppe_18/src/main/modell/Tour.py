import json
import uuid


class Tour:
    # duration in hours
    # cost in dollars

    def __init__(self, title, destination, duration, cost, pictureURL, language="English", max_travelers=15):
        self.pictureURL = pictureURL
        self.language = language
        self.destination = destination
        self.title = title
        self.duration = duration
        self.cost = cost
        self.max_travelers = max_travelers
        self.booked = 0
        self.tour_Id = str(uuid.uuid4())

    def book_tour(self):
        if not self.booked >= self.max_travelers:
            self.booked += 1
            return True
        else:
            return False

    def get_tour_description(self):
        return f"This tour will take you to {self.destination} for {self.duration} hours, and is " \
               f"offered in {self.language}"


