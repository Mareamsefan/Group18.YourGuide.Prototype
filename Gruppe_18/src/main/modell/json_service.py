import json


def save_to_json(entity, filename):
    try:
        with open(filename, "r") as json_file:
            filedata = json.load(json_file)
    except FileNotFoundError:
        filedata = []

    data = entity.to_dict()
    filedata.append(data)

    with open(filename, "w") as json_file:
        json.dump(filedata, json_file, indent=4)


def save_to_stream(entity, io_stream):
    data = entity.to_dict()
    try:
        filedata = json.load(io_stream)
    except json.JSONDecodeError:
        filedata = []

    filedata.append(data)

    io_stream.truncate(0)
    io_stream.seek(0)
    json.dump(filedata, io_stream, indent=4)


def to_dict(entity):
    pass


def from_dict(entity, data):
    pass
