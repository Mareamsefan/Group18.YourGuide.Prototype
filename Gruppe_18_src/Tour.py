import json


class Tour:
    # duration in hours
    # cost in dollars
    def __init__(self, destination, duration, cost, pictureURL, language="English", max_travelers=15):
        self.pictureURL = pictureURL
        self.language = language
        self.destination = destination
        self.duration = duration
        self.cost = cost
        self.max_travelers = max_travelers
        self.booked = 0

    def book_tour(self):
        if not self.booked >= self.max_travelers:
            self.booked += 1
            return True
        else:
            return False


    def get_tour_description(self):
        return f"This tour will take you to {self.destination} for {self.duration}, and is " \
               f"offered in {self.language}"

    def save_to_json(self, filename):

        tour_data = {
            "destination": self.destination,
            "cost": self.cost,
            "duration": self.duration,
            "pictureURL": self.pictureURL,
            "language": self.language,
            "max_travelers": self.max_travelers
        }
        with open(filename, "w") as json_file:
            json.dump(tour_data, json_file, indent=4)


