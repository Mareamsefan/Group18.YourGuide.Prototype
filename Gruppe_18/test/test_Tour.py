from io import StringIO

import pytest
from Gruppe_18.src.main.modell.Tour import Tour
from approvaltests import verify, Options
from approvaltests.scrubbers import scrub_all_guids


approval_options = Options().with_scrubber(scrub_all_guids)

@pytest.fixture
def tour():
    return Tour(
        "Italy",
        4,
        255,
        "https://www.hdwallpaper.nu/wp-content/uploads/2015/05/colosseum-1436103.jpg",
        "English",
        20,
    )

# kan man forenkle assertene her med noen pytest-funksjoner?
def test_if_tour_is_created_and_saved(tour):
    io_stream = StringIO()
    tour.save_to_stream(io_stream)
    saved_data = io_stream.getvalue()
    io_stream.seek(0)
    verify(saved_data, options=approval_options)

def test_book_tour_with_available_space(tour):
    assert tour.book_tour() == True


def test_book_tour_when_fully_booked(tour):
    tour.max_travelers = 1
    tour.booked = 1
    # assert tour.book_tour() == True
    assert tour.book_tour() == False


def test_book_multiple_tours_with_available_space(tour):
    tour.max_travelers = 2
    assert tour.book_tour() == True
    assert tour.book_tour() == True


def test_book_more_tours_than_available_space(tour):
    tour.max_travelers = 2
    tour.booked = 2
    assert tour.book_tour() == False

#approvaltest


def test_tour_description_generation_is_as_expected(tour):
    given_description = tour.get_tour_description()
    expected_description = "This tour will take you to Italy for 4 hours, and is " \
               "offered in English"
    assert given_description == expected_description


def test_tour_description_generation_is_not_as_expected(tour):
    given_description = tour.get_tour_description()
    tour.destination = "Montenegro"
    expected_description = "This tour will take you to Montenegro for 4 hours, and is " \
               "offered in English"
    assert given_description != expected_description


def test_tour_information_is_saved_to_json(tour):
    tour.save_to_json("tour.json")
    assert tour.check_if_tour_is_saved() == True
