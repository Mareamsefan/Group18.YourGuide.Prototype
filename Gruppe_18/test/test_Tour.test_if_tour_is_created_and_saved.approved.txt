[
    {
        "pictureURL": "https://www.hdwallpaper.nu/wp-content/uploads/2015/05/colosseum-1436103.jpg",
        "language": "English",
        "destination": "Italy",
        "title": "A tour through italy",
        "duration": 4,
        "cost": 255,
        "max_travelers": 20,
        "booked": 0,
        "tour_Id": "<guid_0>"
    }
]
